crates/wasmtime/src/runtime/vm/instance.rs

```diff
@@ -35,9 +36,149 @@ use wasmtime_environ::{
+/// the store, so long as you don't do (1-2) again. Note also that the borrow
+/// checker prevents repeating step (3) if you never repeat (1-2). In general,
+/// steps (1-3) should be done in a single, common, internally-unsafe,
+/// plumbing-code bottleneck and the raw pointer should never be exposed to Rust
+/// code that does (4) after the `InstanceAndStore` is created. Follow this
+/// pattern, and everything using the resulting `Instance` and `Store` can be
+/// safe code (at least, with regards to accessing the store itself).
+///
+/// As an illustrative example, the common plumbing code for our various
+/// libcalls performs steps (1-3) before calling into each actual libcall
+/// implementation function that does (4). The plumbing code hides the raw vmctx
+/// pointer and never gives out access to it to the libcall implementation
+/// functions, nor does an `Instance` expose its internal vmctx pointer, which
+/// would allow unsafely repeating steps (1-2).
+#[repr(transparent)]
+pub struct InstanceAndStore {
+    instance: Instance,
+}
+
+impl InstanceAndStore {
+    /// Converts the provided `*mut VMContext` to an `InstanceAndStore`
+    /// reference and calls the provided closure with it.
+    ///
+    /// This method will move the `vmctx` pointer backwards to point to the
+    /// original `Instance` that precedes it. The closure is provided a
+    /// temporary reference to the `InstanceAndStore` with a constrained
+    /// lifetime to ensure that it doesn't accidentally escape.
+    ///
+    /// # Safety
+    ///
+    /// Callers must validate that the `vmctx` pointer is a valid allocation and
+    /// that it's valid to acquire `&mut InstanceAndStore` at this time. For
+    /// example this can't be called twice on the same `VMContext` to get two
+    /// active mutable borrows to the same `InstanceAndStore`.
+    ///
+    /// See also the safety discussion in this type's documentation.
+    #[inline]
+    pub(crate) unsafe fn from_vmctx<R>(
+        vmctx: *mut VMContext,
+        f: impl for<'a> FnOnce(&'a mut Self) -> R,
+    ) -> R {
+        debug_assert!(!vmctx.is_null());
+
+        const _: () = assert!(mem::size_of::<InstanceAndStore>() == mem::size_of::<Instance>());
+        let ptr = vmctx
+            .byte_sub(mem::size_of::<Instance>())
+            .cast::<InstanceAndStore>();
+
+        f(&mut *ptr)
+    }
+
+    /// Get the underlying `dyn VMStore` trait object for this
+    /// `InstanceAndStore`.
+    #[inline]
+    pub(crate) fn store_mut(&mut self) -> &mut (dyn VMStore + 'static) {
+        unsafe { &mut *self.store_ptr() }
+    }
+
+    /// Unpacks this `InstanceAndStore` into its underlying `Instance` and `dyn
+    /// VMStore`.
+    #[inline]
+    pub(crate) fn unpack_mut(&mut self) -> (&mut Instance, &mut (dyn VMStore + 'static)) {
+        unsafe {
+            let store = &mut *self.store_ptr();
+            (&mut self.instance, store)
+        }
+    }
```

Comment:
Can the `'static` here be removed? In theory that's a "lie" because we don't constrain `T: 'static` in the `Store<T>`.

Reply:
Unfortunately, I don't think it can be or else the borrow is invariant and the `InstanceAndStore` is unusable after calling unpack once, which isn't acceptable for `Caller::with`, which needs to access the pair multiple times. It miiight have been an issue in certain libcalls too, I can't quite remember.

Morally, I see this as a "smaller lie" than our current status quo where we just whip up a borrow from a raw pointer whenever we feel like it...

I can try poking at it again, but if I can't figure out a way to remove the `'static`, do you think this is a blocker to landing this?

Reply:
Hm this is surprising. I'm worried about labeling this `'static` when it wasn't since that to me isn't a failure condition we worried about before (I think at least?). Mind trying to remove it and see what happens?

I'm also surprised if `InstanceAndStore` is unusable after one call, as that probably means we have this whole tying typed incorrectly anyway...

Reply:
I guess we could maybe get rid of `InstanceAndStore` and just have

```rust
impl Instance {
    pub(crate) unsafe fn from_vmctx<R>(vmctx: *mut VMContext, f: impl FnOnce(&mut StoreOpaque, &mut Instance) -> R) -> R {
        // ...
    }
}
```

(And another version for `StoreInner<T>`)

Reply:
This would require moving all the `VMStore` trait methods to `StoreOpaque`